ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java

```diff
@@ -76,18 +76,27 @@ public static UnsignedLong compute_slots_since_epoch_start(UnsignedLong slot) {
   /**
    * Get the ancestor of ``block`` with slot number ``slot``.
    *
-   * @param store
+   * @param forkChoiceStrategy
    * @param root
    * @param slot
    * @return
    * @see
    *     <a>https://github.com/ethereum/eth2.0-specs/blob/v0.10.1/specs/phase0/fork-choice.md#get_ancestor</a>
    */
-  private static Bytes32 get_ancestor(ReadOnlyStore store, Bytes32 root, UnsignedLong slot) {
-    BeaconBlock block = store.getBlock(root);
-    if (block.getSlot().compareTo(slot) > 0) {
-      return get_ancestor(store, block.getParent_root(), slot);
-    } else if (block.getSlot().equals(slot)) {
+  private static Bytes32 get_ancestor(
```

Comment:
Rather than throwing `UnknownBlockRootException` which is then caught as a routine result in each place that calls this - would be better to return `Optional` from here.
```
return forkChoiceStrategy
        .blockSlot(root)
        .flatMap(
            blockSlot -> {
              if (blockSlot.compareTo(slot) > 0) {
                return get_ancestor(
                    forkChoiceStrategy,
                    forkChoiceStrategy.blockParentRoot(root).orElseThrow(),
                    slot);
              } else if (blockSlot.equals(slot)) {
                return Optional.of(root);
              } else {
                // root is older than the queried slot, thus a skip slot. Return earliest root prior
                // to slot.
                return Optional.of(root);
              }
            });
```

Reply:
Makes sense. Done.